features/285-design-webhooks/design.md

```diff
@@ -0,0 +1,129 @@
+# next webhooks design
+
+## Overview
+
+skygear sends webhooks for events that happen in your app, and are designed to help a developer to monitor events, sync database, and many other use cases.
+
+When one of events is triggered, skygear sends a HTTP POST with a payload to the webhook's configured URL.
+
+## SYNC and ASYNC hooks
+
+There are two types of webhook:
+
+- `SYNC` - gear's operation will be blocked until a response is received. Gear should respect `SYNC` hook's response, if it indicates failed, operation should abort and return a formatted error response to the client.
+- `ASYNC` - gear's operation won't be blocked and the response from the webhook is ignored.
+
+## Request Headers
+
+Following headers will be added to webhook POST request:
+
+- `X-Skygear-Webhook-Req-ID`: an ID of the request.
+- `X-Skygear-Webhook-Async`: if is a `async` hook, then the value is `TRUE` otherwise is `FALSE`. 
+- `X-Skygear-Webhhok-Signature`: a signature string of the webhook event, a develop should can verify the request by skygear's SDK or manually.
+
+## Request Payload
+
+Each event type has a payload which carries relevant information, and each payload data is defined by each gear.
+
+All event payloads follows following payload format:
+
+```json=
+{
+    "event": "event_name",
+    "data": {
+        "key_name_1": "event_data_key_value",
+        "key_name_2": "event_data_key_value",
+        ...,
+        "context": {
+            "user": <current user object>,
+            "req": {
+                "path": "request_path",
+                "body": <original_request_body>,
+                "id": "request_id"
+            }
+        }
+    }
+}
+```
```

Comment:
I prefer `context` to be outside the `data` dictionary because

- conceptually the boundary between `data` and `context` seems clear
- `context` is a common thing to most hooks
- `data` will not be able to use the key `context`

Reply:
Understand, I had moved `context` to the root level.